server/chat-plugins/lottery.js

```diff
@@ -0,0 +1,248 @@
+}
+if (!lotteries || typeof lotteries !== 'object') lotteries = {};
+
+class Lottery {
+	/**
+     * @param {string} roomid
+     */
+	constructor(roomid) {
+		this.roomid = roomid;
+	}
+	/**
+	 * @param {number} maxWinners
+	 * @param {string} name
+	 * @param {string} markup
+	 */
+	create(maxWinners, name, markup) {
+		if (!lotteries[this.roomid]) {
+			lotteries[this.roomid] = {maxWinners, name, markup, participants: {}};
+			this.write();
+		}
+	}
+	write() {
+		FS(LOTTERY_FILE).writeUpdate(() => JSON.stringify(lotteries));
+	}
+	destroy() {
+		delete lotteries[this.roomid];
+		this.write();
+	}
+	/**
+     * @param {User} user
+     */
+	addUser(user) {
+		if (!lotteries[this.roomid].participants[user.latestIp]) {
+			lotteries[this.roomid].participants[user.latestIp] = user.name;
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	/**
+     * @param {User} user
+     */
+	removeUser(user) {
+		if (lotteries[this.roomid].participants[user.latestIp]) {
+			delete lotteries[this.roomid].participants[user.latestIp];
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	getWinner() {
+		const winners = [];
+		const lotteryData = lotteries[this.roomid];
+		let participantsKeys = Object.keys(lotteryData.participants);
+		for (let i = 0; i < lotteryData.maxWinners; i++) {
+			const randomKey = participantsKeys[participantsKeys.length * Math.random() << 0];
+			const winner = lotteryData.participants[randomKey];
+			winners.push(winner);
+			participantsKeys = participantsKeys.filter(username => username !== winner);
+		}
+		return winners;
+	}
+}
+
+/** @type {ChatCommands} */
+exports.commands = {
+	lottery: {
+		'!': true,
+		''(target, room) {
+			if (!room) {
+				return this.errorReply("This can only be used in rooms.");
+			}
+			const lotteryData = lotteries[room.id];
+			if (!lotteryData) {
+				return this.errorReply("This room doesn't have a lottery running.");
+			}
+			return this.parse(`/join view-lottery-${room.id}`);
+		},
+		'!create': true,
+		create(target, room, user) {
+			if (!this.can('declare', null, room)) return;
+			if (!room) {
+				return this.errorReply("This can only be used in rooms.");
+			}
+			if (lotteries[room.id]) {
+				return this.errorReply(`There is already a lottery running.`);
+			}
+			const lottery = new Lottery(room.id);
+			let [maxWinners, name, ...markup] = target.split(',').map(val => val.trim());
+			if (!(maxWinners && name && markup.length)) {
+				return this.errorReply("You're missing a command parameter - see /help roomevents for this command's syntax.");
+			}
+			const maxWinnersNum = parseInt(maxWinners);
+			if (isNaN(maxWinnersNum)) {
+				return this.errorReply(`${maxWinners} is not a valid number.`);
+			}
+			if (maxWinnersNum < 1) {
+				return this.errorReply('The maximum winners should be at least 1.');
+			}
+			lottery.create(maxWinnersNum, name, markup.join(','));
+			this.sendReply('The lottery was successfully created.');
```

Comment:
Considering this sends a declare-like message right after with the user's username, I wonder if this reply is necessary or not

Reply:
I feel like feedback other than the message the is sent to everybody is nice.

Reply:
If the consensus is against me,  though I'll remove it, I don't really mind that much.